orttraining/orttraining/core/optimizer/lora_conv1d_replacement.cc

```diff
@@ -0,0 +1,141 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/initializer.h"
+#include "orttraining/core/optimizer/lora_conv1d_replacement.h"
+#include "core/graph/graph_utils.h"
+
+/*
+  in LoRA code, it will use conv1d to do projection for qkv,
+  while the conv1d calculation is mathematically equivalent to matmul, and matmul is much faster than conv1d.
+  the graph transformation is doing the following graph substitution:
+  1. the input graph is:
+  conv_input  conv_weight
+        \       /
+         \     /
+          conv1d
+
+  2. the output graph is as follows,
+     the number of matmul is equal to attribute "group" of conv1d
+    conv_input   conv1d.group    conv_weight  conv1d.group
+      \          /                   \         /
+       \        /                   Squeeze   /
+        \      /                       \     /
+          Split                        Split
+          /...\                        /...\
+          ||   |_____________Matmul___||___|
+          ||                   /      ||
+          ||_________Matmul___/_______||
+                       /     /
+                      /...../
+                      Concat
+*/
+namespace onnxruntime {
+bool NodeCanBeReplacedByMatmul(const Node& node) {
+  // if node type is Conv, and attr "dilations" is 1, "kernel_shape" is 1, "stride" is 1, then it can be replaced by matmul
+  // kernel_shape is 1 means it is conv1d
+  if (!graph_utils::IsSupportedOptypeVersionAndDomain(node, "Conv", {1, 11})) {
+    return false;
+  }
+  const auto* dilations = graph_utils::GetNodeAttribute(node, "dilations");
+  const auto* kernel_shape = graph_utils::GetNodeAttribute(node, "kernel_shape");
+  const auto* stride = graph_utils::GetNodeAttribute(node, "strides");
+  if (dilations == nullptr || kernel_shape == nullptr || stride == nullptr) {
+    return false;
+  }
+  if ((dilations->ints_size() && dilations->ints(0) != 1) || (kernel_shape->ints_size() && kernel_shape->ints(0) != 1) || (stride->ints_size() && stride->ints(0) != 1)) {
+    return false;
+  }
+
+  return true;
+}
+
+void Conv1dToMatmul(Graph& graph, Node& conv) {
+  // shape of conv1d input: [batch_size, in_channels, in_length]
+  // shape of conv1d weight:[output_channels, input_channels/group, kernel_shape], kernel_shape is 1
+  // we need to split the input into "group", and squeeze&split the weight, and then do matmul
+  auto node_description = std::string("LoRAConv1dReplacement");
```

Comment:
nit: better to use explicit type if possible, making it easier to inspect when revisiting the code.